File: app/src/main/java/rz/com/catolico/adapter/GenericAdapter.java
package rz.com.catolico.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

public abstract class GenericAdapter<T> extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private Context context;
    private List<T> mItems;

    public abstract RecyclerView.ViewHolder setupViewHolder(ViewGroup parent, int viewType);

    public abstract void onBindData(RecyclerView.ViewHolder holder, final T item);

    public GenericAdapter(Context context , List<T> mItems) {
        this.context = context;
        this.mItems = mItems != null ? mItems : new ArrayList<>();
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        RecyclerView.ViewHolder holder = setupViewHolder(parent, viewType);
        return holder;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        onBindData(holder, this.mItems.get(position));
    }

    @Override
    public int getItemCount() {
        return this.mItems.size();
    }

    public T getItem(int position) {
        return this.mItems.get(position);
    }

    public Context getContext() {
        return this.context;
    }

}

File: app/src/main/java/rz/com/catolico/adapter/GenericAdapter.kt
package rz.com.catolico.adapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.ViewGroup
import rz.com.catolico.utils.ActivityUtils.Companion.isUserLogged

import java.util.ArrayList

abstract class GenericAdapter<T>(val context: Context, mItems: List<T>?) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val mItems: List<T> = mItems ?: ArrayList()
    protected var isUserLogged : Boolean = isUserLogged(context)

    abstract fun setupViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder

    abstract fun onBindData(holder: RecyclerView.ViewHolder, genericType: T)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return setupViewHolder(parent, viewType)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        onBindData(holder, this.mItems[position])
    }

    override fun getItemCount(): Int {
        return this.mItems.size
    }

    fun getItem(position: Int): T {
        return this.mItems[position]
    }

}